### PR TITLE
feat(coremlit/lid): a backend-neutral spoken-language-identification door (#100)

### LINKED ISSUE #100: audio LID: speechbrain lang-id-voxlingua107-ecapa on CoreML
User-ordered (2026-08-28), companion to #98/#99: coremlit grows a standalone audio language-identification family — **speechbrain/lang-id-voxlingua107-ecapa** (ECAPA-TDNN over VoxLingua107; 107 languages).

## Why
mediagraph's ASR road currently gets language only from whisper's own argmax probe (top-1, weak). The recorded optimization (findit-studio/mediagraph#95) is a combined vad_lid node emitting per-speech-span language evidence — blocked solely on coremlit having a real LID door. This model is that door: per-span top-k language posteriors with confidences, which is exactly the `Vec<DetectedLanguage>` shape the graph stores.

## Suggested shape (house patterns)
- Module `audio::lid`, feature `lid`, sans-I/O door at 16 kHz mono `&[f32]` (the audio towers' law): `identify(&[f32]) -> Result<Vec<(code, f32)>>` (top-k, sorted), model `Send + !Sync`, load+prewarm at construction.
- Front end: ECAPA takes log-mel fbanks (60-dim, 25 ms / 10 ms hop at 16 kHz in the speechbrain recipe) — a Rust front end beside the CED/whisper mel implementations, goldens against the speechbrain reference outputs (the transformers-fp32-golden discipline the siglip port used).
- CoreML export of the ECAPA graph + classifier head; artifact layout per the house (tier dir with .mlmodelc + any sidecar); the 107-label roster ships as data with a pinned identity (the tokenizer-SHA discipline).
- Labels are ISO-639-ish codes — the downstream maps them onto ingraph's Language family (which folds aliases itself), so raw codes are fine; just keep the roster exact and stable.

## Consumer contract (for sizing)
One call per speech span (VAD upstream cuts spans; spans are seconds-long). Top-k (k≈3–5) with confidences beats top-1. Batch door not needed — same finding as the vision tower (mediagraph#93's census posture).

### PR DESCRIPTION
**Pull request by uqio**
Closes #100. `audio::lid` behind feature `lid` — sans-I/O 16 kHz mono `&[f32]` in, top-k language scores out. The public shape names neither VoxLingua nor ECAPA, because #101 (3D-Speaker) and #102 (FireRedLID) plug into this same seam.

Backend: `aufklarer/SpeechBrain-ECAPA-VoxLingua107-21M-CoreML` @ `2aa4d715`, Apache-2.0, 21,246,443 params / 40.77 MiB.

## The trap this port exists to avoid

SpeechBrain's `InputNormalization(norm_type=sentence)` is **already fused into the CoreML graph** — `model.mil`'s `func main` opens with `reduce_mean(axes=[1], keep_dims=true)` then `sub`. Upstream's `frontend.py` ships a `sentence_mean_normalize()` helper that its own README example deliberately never calls.

**A faithful port of SpeechBrain's *pipeline* double-subtracts the mean and is silently wrong** — right shape, plausible output, degraded accuracy. The block comment sits at the exact line a reader would add it: after the final `top_db` write loop, immediately before `Ok(())`.

## Four front-end conventions a generic mel implementation gets wrong

| | SpeechBrain | the usual assumption |
|---|---|---|
| window | **periodic Hamming** | Hann |
| centre pad | **constant zero** (overrides torch's default) | reflect |
| triangles | **symmetric, `band[i] = hz[i+1] − hz[i]` both sides, no area norm** | Slaney/HTK normalised |
| floor | **global `top_db = 80`** over time *and* mel | per-frame |

Each has a test that fails under the wrong convention, verified by an **8-mutation sweep**: transposed write reds 4 tests including the layout killer; Hann 2; symmetric window 2; reflect padding 2; asymmetric triangles 2; Slaney norm 2; per-frame `top_db` 1; f64 tables red the filterbank anchor. Source confirmed byte-identical after restore.

## On the two SHA anchors

**The filterbank reproduces bit-exactly** (`264347e3…`, all 12,060 f32 weights) — and only with f32 table arithmetic; an f64 build narrowed at the end moves 380 of them.

**The window anchor does not, and cannot portably.** `ed725a5a…` encodes **NumPy's vectorized float32 cosine, which is not correctly rounded** — it disagrees with the exactly-rounded f32 cosine at 66 of 400 positions (checked against a 60-digit series evaluation). No natural Rust formulation reaches it; three were tried, in Rust and Python, and they agree with each other and not with NumPy.

So our own digest is pinned as the regression anchor, the reference digest and the reasoning are recorded in the test, and the real content moved into invariants no ulp can move: `w[0] = 0.54 − 0.46` exactly, `w[399] ≈ 0.080057 > w[0]` (periodic, not symmetric), sum `216.0` against the symmetric window's `216.46`.

**Stronger evidence than either anchor:** the front end reproduces the reference mel end-to-end at **max 1.9e-5 dB / RMSE 1.3e-6 dB** over a 63 dB range, and the full pipeline reproduces the upstream Thai anchor's top-1 log-probability to **1.1e-7** (`−0.010064`, index 94 `th`), with top-3 codes and values matching the Swift probe.

## Contract

```
input   mel_features      fp32 [1, frames, 60]   frames 10..=3001   TIME-major
output  log_probabilities fp32 [1, 107]          natural log, already softmaxed (Σexp = 1.0)
```

Note the layout: ced writes freq-major, this model wants time-major. A transposed write still yields a valid-shaped tensor, hence the dedicated killer — and it asserts *every* first-half frame sits on the global floor and *every* second-half frame is >30 dB above it, because a mean-vs-mean form would survive the transpose (offset `60·t` cycles through loud content).

`identify` returns `Vec<LanguageScore>` — `&'static Language` plus `log_probability: f32` — not `(usize, f32)`. A bare tuple gives no way to distinguish a natural-log probability from a logit or a probability, and **both wrong readings produce numbers rather than errors**; a raw index invites lookup against the wrong roster. Same shape as ced's `EventPrediction`.

Frames outside `10..=3001` (1440..=480159 samples, 0.0900–30.0099 s) are a typed error raised **before** the model is called, never the runtime's message. Windowing >30 s is deliberately left to a follow-up.

## Compute placement — measured

| computeUnits | load | 13 s clip | 3 s clip |
|---|---|---|---|
| `All` | 113 ms | 13.9 ms | 4.8 ms |
| `CpuAndGpu` | 50–100 ms | 13.8 ms | 7.2 ms |
| `CpuOnly` | 21 ms | 24.7 ms | 6.7 ms |
| `CpuAndNeuralEngine` | **2440–2816 ms** | **145 ms** | **36 ms** |

The ANE arm is pathological and additionally emits `BNNS Graph Shape Deduction: Unsupported kernel id 512`. `DEFAULT_COMPUTE = All` is bit-identical to `CpuAndGpu`, i.e. it dispatches to the GPU — so the house default is **correct by luck, not by design**. The doc says so, names the OS-change risk, and `default_placement_agrees_with_the_gpu_arm_bit_for_bit` is the gate that watches it.

## Labels

107-entry roster committed and pinned the way `align/vocab` pins its tokenizer: `include_bytes!`, a size const, embedded-vs-committed byte equality, and semantic index pins. Byte-identical to the author's MLX export and to upstream's `label_encoder.txt` in index order, zero mismatches. **Legacy codes `iw` (44) and `jw` (46) are preserved deliberately** — upstream keeps them so indices never shift; downstream alias folding must handle them.

## Deliberately unwired

`ci.yml` / `coverage.yml` matrices, `MODELS_LOCK`, `FEATURE_MAP.md`, `README`, `CHANGELOG`. Registration is an 11-item checklist with a tripwire (`feature_map.rs` holds a verbatim YAML replica of ci.yml as its own parser fixture) and deserves its own change. The one exception is `feature_map.rs`'s manifest-mirror row — without it `feature_names_match_the_pinned_set` reds and the tree ships broken.

## Verify

`fmt=0 clippy=0 lib_lid=0 lib_all=0 doc=0 default_norun=0 all_targets=0 rustdoc=0` — lib 1280 passed / 0 failed. **All 9 model gates ran green against the staged artifact**, not merely compiled.

### CODE REVIEW

**Comment by uqio** [comment]
## Codex adversarial review: **APPROVE** (R1, no material findings)

> Ship: no substantive product-code defect with a concrete triggering input was found in the permitted diff.

Clean on the first round, under a product-code-only fence (tests, docs, changelog and CI excluded at every severity). The review was aimed specifically at the failure modes that produce **plausible numbers rather than errors**, since those are the ones that survive testing:

- any normalization in the front end that would compose with the graph's **fused sentence-mean subtraction** — the trap that would silently halve accuracy;
- a **transposed or mis-strided write** that still yields a valid-shaped tensor, given this model is time-major where the sibling `ced` door is freq-major;
- a second `exp`/`log`/softmax applied to output that is *already* log-softmaxed;
- an off-by-one into the 107-label roster;
- panics, overflow, non-finite propagation and unbounded allocation anywhere in the hand-rolled DSP, including subnormals, infinities and NaN.

None found.

Worth recording why this landed clean where a comparable door took five rounds: **the artifact was censused before a line was written.** The fused normalization, the four non-standard front-end conventions, the `[10, 3001]` frame range, the log-probability output, the time-major layout and the pathological ANE arm were all established from the model itself — so they became tests rather than review findings.

Still deliberately unwired, and tracked separately: `ci.yml` / `coverage.yml` matrices, `MODELS_LOCK`, `FEATURE_MAP.md`, `README`, `CHANGELOG`.
